### PR TITLE
Implemented Required Features

### LINKED ISSUE #1: Image Histogram
## Summary of the request

Create a new handle that allows picking one image from a list and returns the histogram of an image. The histogram can be rendered either in the frontend (JavaScript library, for extra points), or in the backend and passed as an image (python / matplotlib).
**This is a new feature. Old functionalities must be preserved.**
Difficulty: 
image/image(If using Javascript)

## Subtasks
* Copy the classification API form and modify it where it is required.
* Implement the new API that takes as input the id of an image and returns the histogram.
* Visualize the histogram and the image in the GUI. 

## Plan for the implementation

Modify the code such that the new interface shows the dropdown menu to select the image and the submit button.
It is not required to use queues. The API can be implemented as showing the form if a GET request is issued, and the results if the POST is validated.

image

Useful links: 
* https://stackoverflow.com/questions/22159160/python-calculate-histogram-of-image

### LINKED ISSUE #2: Image transformation
## Summary of the request

Create a handler that allows picking an image and set parameters for an image transformation (https://pillow.readthedocs.io/en/stable/reference/ImageEnhance.html). The user should be allowed to specify the color, brightness, contrast, and sharpness to apply to the original image. The transformed image should appear in the output interface.
**This is a new feature. Old functionalities must be preserved.**
Difficulty:
image

## Subtasks

* Copy the classification API interface.
* Create a form for selecting the transformation.
* Implement the API that applies the transformation to the selected image.
* Implement the visualization of results.

## Plan for the implementation

Modify the code such that the new interface shows the dropdown menu to select the image, the submit button, and a NEW field that allows specifying values for the transformation to apply. The values for the transformation should have default values, but should be editable by the user. Use Flask forms to create the interface for selecting the transformation. Use correct field types for the inputs.
It is not required to use queues. The API can be implemented as showing the form if a GET request is issued, and the results if the POST is validated.

Useful links:
* https://pillow.readthedocs.io/en/stable/reference/ImageEnhance.html
* [Flask form fields](https://wtforms.readthedocs.io/en/2.3.x/fields/)

### LINKED ISSUE #3: Download results buttons
## Summary of the request

Add the “download results” button to the web interface. The button should allow the user to download the results (classification scores) as a JSON file and as a png file showing the top 5 scores in a plot (bar chart).
**All the functionalities must be preserved, this is just a new feature**.
Difficulty:
image

## Subtasks

* Add the "download results" button to the interface.
* Add the "download plot" button to the interface.
* Implement the API for returning a JSON file with the results.
* Implement the API for returning the plot.
* Connect the buttons with the APIs.

## Plan for the implementation

The easiest way to implement the new functionality is to create the files on the go when the button is pressed, through the button URL linked with an HTML tag (```<a href="URL">```, see references below).

image

Useful links:
* https://www.w3schools.com/tags/tag_a.asp
* https://stackoverflow.com/questions/30024948/flask-download-a-csv-file-on-clicking-a-button

### LINKED ISSUE #4: Upload Image Button
## Summary of the request

Add the “upload image” functionality. Implement a new route that instead of showing the dropdown list of images to select, allows the user to select an image from the computer, and returns the classification results.
The button should allow the user to upload a custom image and classify it. The image should be displayed in the output interface along with the classification scores.
**The old functionalities should be preserved, this is an additional feature**.
Difficulty:
image

## Subtasks

* Copy the route and HTML file of the classification API.
* Add the new route to the routes init file.
* Add "upload" button in the frontend (HTML).
* Implement the API for storing the image.
* Pass the image to the classification API.

## Plan for the implementation

image

Implement using the Flask APIs and modifying a copy of the available classification API. No particular additional library is required.

Useful links:
* https://flask.palletsprojects.com/en/2.0.x/patterns/fileuploads/

### PR DESCRIPTION
**Pull request by christianscano**
## Changelog

* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

  - __Image Histogram__
Feature which allows a user to choose an image from a dateset and see its RGB histogram (implemented via Javascript)

  - __Image Transformation__
 Feature which allows a user to choose an image from a dataset and modify it using the following parameters: _color_, _brightness_,  _contrast_, _sharpness_

  - __Download Results Buttons__
  Feature which allows a user to download the result of a classification in PNG format (Plot) or JSON format (raw data)

  - __Upload Image Button__
  Features which allows a user to upload own image and classify it. Some security checks are added in order to avoid a user from being able to upload arbitrary files. Only PNG, JPEG, JPG formars are allowed.

* **Other information**:(does the pr fix some issues? Tag them with #<issue id>)

  Fixes #1, #2 , #3, #4

### CODE REVIEW

**Comment by christianscano** [comment]
Hi @maurapintor github doesn't allow me to add you as a reviewer

File: app/routes/classifications_upload.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+                        "model_id": model_id,
+                        "img_id": image_id
+                    },
+                    on_failure=cleanup_temp_image,
+                    on_success=cleanup_temp_image
+                )
+                task = q.enqueue_job(job)
+
+                # returns the image classification output from the specified model
+                # return render_template('classification_output.html', image_id=image_id, results=result_dict)
```
Comment (maurapintor) [inline app/routes/classifications_upload.py]: Can you remove these commented lines here?

**Review by maurapintor — CHANGES_REQUESTED** [review]
Overall, outstanding implementation and highly-readable code! 👍 

There is just a commented line of old code to cleanup. After you fix that, I will accept the PR.

File: app/routes/classifications_upload.py (hunk 1 above)
Comment (christianscano) [reply app/routes/classifications_upload.py]: Done

**Review by maurapintor — APPROVED** [review]
LGTM, merging 😎

**Comment by maurapintor** [comment]
fixes #2, #3, #4 as well

**Pull request merged**